### DISCUSSION TITLE
Nested rc_clauses parsing problems

### DISCUSSION THREAD

**Discussion by Fxlix** [explosion/spaCy#8249] [Help: Coding & Implementations]
When parsing german Sentences with nested rc_clauses ("Die schöne Frau, die Fritten, welche schwarz sind, isst, und die nervige Katze baden im Fluss." ) the spaCy Dependency parser pipeline from the de_dep_news_trf Model  will fail with high propability.
When recursively cutting the innermost "rc_clause", the Dependency parser does what it is expected to do. 
So I thought I could reparse the Sentence without the rc_clause and then line up the cutted rc_clause via retokenizer splitting. 
Computing the new heads and attrs should work fine - But the retokenizer just prints:
"ValueError: [E117] The newly split tokens must match the text of the original token. New orths: Fritten,welcheschwarzsind,. Old text: Fritten""

I understand that inserting and changing tokens is not what spacy is intended for, because its intended to keep the origin, but I really could need a good workaround, because its a recurring error. 

## Your Environment

- **spaCy version:** 3.0.6
- **Platform:** Linux-5.4.72-microsoft-standard-WSL2-x86_64-with-debian-10.9
- **Python version:** 3.7.3
- **Pipelines:** de_dep_news_trf (3.0.0), de_core_news_lg (3.0.0)

**Comment by svlandeg** [comment]
Hi!

This approach does seem risky to me, but I understand you want to have a workaround for these inaccurate cases.

Can you post a minimal code snippet that showcases exactly what you're trying to do? It will be helpful to try and reproduce the `ValueError` you run into - perhaps we can help fix it.

**Reply by Fxlix** [reply]
Well the ValueError is what results from:
"When splitting tokens, the subtoken texts always have to match the original token text – or, put differently "".join(subtokens) == token.text always needs to hold true. If this wasn’t the case, splitting tokens could easily end up producing confusing and unexpected results that would contradict spaCy’s non-destructive tokenization policy." - SpaCy documentation. 
So theres no need to reproduce the error. 

If a rc-Dependency is found i try to grab the whole subtree, from that token. I iterate through that subtree to check if its interrupted by a comma (Assumption no Conjuncts in that relative Clause).
When its not interrupted its the innermost rc_clause. This span should have the correct Dependencies. Outside of this Span the Dependencies are broken.  So i want to cut it, but keep the POS/STTS-Tags deps and heads from the cutted tokens. 
Then i would like to reparse the sentence without the cutted rc_clause. Now all Dependencies should be correct. 
But now I want to insert the cutted rc_tree in the reparsed Document.

The Code attached isn't the Code I use, because its actually nested inside a Docker container app.py script. But it shows a way to reproduce the "Error"-Not an real Error, because spacy is supposed to do that. 

```

import spacy 
import numpy as np
nlp = spacy.load("de_dep_news_trf")
text ="Die schöne Frau, die Fritten, welche schwarz sind, isst, und die nervige Katze treiben tot im Fluss."

doc = nlp(text)
rc_tree = []
sub_head = -1
has_rc = 0
is_interrupted=  0
left_comma = -1
right_comma = -1
for token in doc:
	if token.dep_ == "rc":
		has_rc =  1
		sub_head = token.head.i
		for child in token.subtree:
			if child.text == ",":
				is_interrupted = 1
		if is_interrupted == 0:
			rc_tree = list(token.subtree)
			left_comma = rc_tree[0].i-1
			right_comma = rc_tree[-1].i+1
			break 

if has_rc == 1 and is_interrupted == 0:
	new_string = ""
	sub_ids = [token.i for token in rc_tree]
	sub_ids_rel = [token.i - sub_head for token in rc_tree]
	sub_dep_ids_rel = [token.head.i - sub_head for token in rc_tree]
	for token in doc:
		if token.i not in sub_ids and token.i != right_comma and token.i != left_comma:
			new_string = new_string + token.text + token.whitespace_
		if token.i == right_comma:
			new_string = new_string + " "

	arr = [sub.text for sub in rc_tree]
	text_arr = np.concatenate(([doc[sub_head].text],[","],arr,[","]))
	attrs = {}
	POS_list = ([token.pos_ for token in rc_tree])
	POS_list.append("PUNCT") #right_comma
	POS_list.insert(0,"PUNCT") # left_comma
	attrs["POS"] =  POS_list
	DEP_list = ([token.dep_ for token in rc_tree])
	DEP_list.append("punct")#right_comma
	DEP_list.insert(0,"punct")# left_comma
	attrs["DEP"] = 	DEP_list

	doc2 = nlp(new_string)
	heads = np.concatenate(( [doc2[sub_head]], [(doc2[sub_head] , sub_i ) for sub_i in sub_dep_ids_rel] ,[doc2[sub_head] ,doc2[sub_head]]),axis=None) 
	
	with doc2.retokenize() as retokenizer:
		retokenizer.split(doc2[sub_head], text_arr,heads=heads,attrs=attrs)


```
